Refactor lookalike character sanitization to use the new type

charliewolfe opened the pull request:
#### 3574f7790d2190267a301c2e211195f2786f6cff
<pre>
Refactor lookalike character sanitization to use the new type
<a href="https://bugs.webkit.org/show_bug.cgi?id=251902">https://bugs.webkit.org/show_bug.cgi?id=251902</a>
rdar://105160778

Reviewed by Wenson Hsieh.

This patch makes changes to lookalike character sanitization such that it
uses the `LookalikeCharacterSanitizationData` struct instead of WTF::String.

* Source/WebCore/page/LookalikeCharactersSanitizationData.h:
(WebCore::LookalikeCharactersSanitizationData::LookalikeCharactersSanitizationData):
(WebCore::LookalikeCharactersSanitizationData::lookalikeCharacters):
* Source/WebKit/Platform/cocoa/NetworkConnectionIntegrityHelpers.h:
(WebKit::LookalikeCharacters::cachedStrings const):
* Source/WebKit/Shared/WebPageCreationParameters.cpp:
(WebKit::WebPageCreationParameters::decode):
* Source/WebKit/Shared/WebPageCreationParameters.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::setLookalikeCharacterStrings):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:

Canonical link: <a href="https://commits.webkit.org/261937@main">https://commits.webkit.org/261937@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f49683ae45953c51e7ccea07c61ec8a0177235bb

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/107551 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/16609 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/40454 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/116713 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/116130 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/111447 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/18036 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/7927 "Built successfully") | [⏳ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/WPE-WK2-Tests-EWS "Waiting to run tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/113310 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/13622 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/96811 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/99719 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/95531 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/28437 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/41285 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/9612 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/29790 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/83077 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/10272 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/6686 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/15769 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/49371 "Passed tests") | | 
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/8346 "Build is in progress. Recent messages:OS: Monterey (12.6.3), Xcode: 13.4.1; Checked out pull request; Reviewed by Wenson Hsieh") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/11833 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

whsieh commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -8399,12 +8399,15 @@ bool WebPage::isUsingUISideCompositing() const
+    for (auto& data : strings) {
+        if (!m_lookalikeCharacterStrings.contains(data.domain))
+            m_lookalikeCharacterStrings.add(data.domain, HashSet<String> { });
+        m_lookalikeCharacterStrings.get(data.domain).add(data.lookalikeCharacters);
```
There's a bit of redundant hashing here — perhaps we could use `HashMap::ensure()` instead, passing in a lambda that returns a new HashSet and then call `add` on the new or extant set, using the `AddResult` iterator.

whsieh commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -8399,12 +8399,15 @@ bool WebPage::isUsingUISideCompositing() const
     m_lookalikeCharacterStrings.reserveInitialCapacity(strings.size());
-    for (auto& string : strings)
-        m_lookalikeCharacterStrings.add(string);
+    for (auto& data : strings) {
```
We should also double check that `data.domain` always represents a valid hash key here, to avoid hash table corruption (e.g. by trying to add a null string which represents the empty value).

whsieh commented on Source/WebCore/page/LookalikeCharactersSanitizationData.h:
```diff
@@ -46,6 +46,12 @@ struct LookalikeCharactersSanitizationData {
     }
 
+    LookalikeCharactersSanitizationData(const String& lookalikeCharacters)
+        : domain({ })
```
Nit - I _think_ you can also omit this initialization here, since it would use the `{ }` initializer by default.

charliewolfe committed:
Refactor lookalike character sanitization to use the new type
https://bugs.webkit.org/show_bug.cgi?id=251902
rdar://105160778

Reviewed by Wenson Hsieh.

This patch makes changes to lookalike character sanitization such that it
uses the `LookalikeCharacterSanitizationData` struct instead of WTF::String.

* Source/WebCore/page/LookalikeCharactersSanitizationData.h:
(WebCore::LookalikeCharactersSanitizationData::LookalikeCharactersSanitizationData):
(WebCore::LookalikeCharactersSanitizationData::lookalikeCharacters):
* Source/WebKit/Platform/cocoa/NetworkConnectionIntegrityHelpers.h:
(WebKit::LookalikeCharacters::cachedStrings const):
* Source/WebKit/Shared/WebPageCreationParameters.cpp:
(WebKit::WebPageCreationParameters::decode):
* Source/WebKit/Shared/WebPageCreationParameters.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::setLookalikeCharacterStrings):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:

Canonical link: https://commits.webkit.org/261937@main
